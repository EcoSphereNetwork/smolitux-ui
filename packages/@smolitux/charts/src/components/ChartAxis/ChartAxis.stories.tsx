import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { ChartAxis } from './ChartAxis';

const meta: Meta<typeof ChartAxis> = {
  title: 'Charts/ChartAxis',
  component: ChartAxis,
  parameters: { layout: 'centered' },
  tags: ['autodocs'],
  argTypes: {
    orientation: { control: { type: 'radio' }, options: ['horizontal', 'vertical'] },
    length: { control: { type: 'number' } },
  },
};
export default meta;

type Story = StoryObj<typeof ChartAxis>;

export const Horizontal: Story = {
  render: (args) => (
    <svg width={200} height={60} viewBox="0 0 200 60">
      <ChartAxis {...args} />
    </svg>
  ),
  args: {
    length: 180,
    ticks: [
      { value: 0, position: 0 },
      { value: 50, position: 0.5 },
      { value: 100, position: 1 },
    ],
    orientation: 'horizontal',
    axisLabel: 'X-Axis',
  },
};

export const Vertical: Story = {
  render: (args) => (
    <svg width={80} height={200} viewBox="0 0 80 200">
      <ChartAxis {...args} />
    </svg>
  ),
  args: {
    length: 180,
    orientation: 'vertical',
    ticks: [
      { value: 0, position: 0 },
      { value: 50, position: 0.5 },
      { value: 100, position: 1 },
    ],
    axisLabel: 'Y-Axis',
  },
};
