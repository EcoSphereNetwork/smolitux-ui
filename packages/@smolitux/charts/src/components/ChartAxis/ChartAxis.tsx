// TODO: Tests fehlen
import React, { forwardRef } from 'react';

export interface ChartAxisTick {
  /** Wert des Ticks */
  value: string | number;
  /** Position relativ (0-1) */
  position: number;
  /** Optionales Label */
  label?: string;
}

export interface ChartAxisProps extends React.SVGAttributes<SVGGElement> {
  /** Orientierung der Achse */
  orientation?: 'horizontal' | 'vertical';
  /** Gesamtlänge der Achse */
  length: number;
  /** Ticks für die Achse */
  ticks?: ChartAxisTick[];
  /** Achsenlabel */
  axisLabel?: string;
  /** Abstand der Ticks */
  tickSize?: number;
}

/**
 * Universelle Achsen-Komponente für SVG-Charts
 */
export const ChartAxis = forwardRef<SVGGElement, ChartAxisProps>(
  (
    {
      orientation = 'horizontal',
      length,
      ticks = [],
      axisLabel,
      tickSize = 6,
      className = '',
      ...rest
    },
    ref,
  ) => {
    const isHorizontal = orientation === 'horizontal';
    const axisLine = isHorizontal
      ? { x1: 0, y1: 0, x2: length, y2: 0 }
      : { x1: 0, y1: 0, x2: 0, y2: length };

    return (
      <g ref={ref} className={`smolitux-chart-axis ${className}`.trim()} {...rest} data-testid="chart-axis">
        <line {...axisLine} stroke="currentColor" />
        {ticks.map((tick) => {
          const pos = tick.position * length;
          const tickProps = isHorizontal
            ? { x1: pos, y1: 0, x2: pos, y2: tickSize }
            : { x1: 0, y1: pos, x2: -tickSize, y2: pos };
          const textProps = isHorizontal
            ? { x: pos, y: tickSize + 12, textAnchor: 'middle' as const }
            : { x: -tickSize - 2, y: pos, textAnchor: 'end' as const, dominantBaseline: 'middle' as const };
          return (
            <g key={tick.value} className="tick">
              <line {...tickProps} stroke="currentColor" />
              <text {...textProps} fontSize="12" fill="currentColor">
                {tick.label ?? tick.value}
              </text>
            </g>
          );
        })}
        {axisLabel && (
          <text
            x={isHorizontal ? length / 2 : -tickSize - 20}
            y={isHorizontal ? tickSize + 28 : length / 2}
            textAnchor="middle"
            transform={isHorizontal ? undefined : `rotate(-90, ${-tickSize - 20}, ${length / 2})`}
            fontSize="12"
            fontWeight="bold"
            fill="currentColor"
          >
            {axisLabel}
          </text>
        )}
      </g>
    );
  },
);

ChartAxis.displayName = 'ChartAxis';

export default ChartAxis;
