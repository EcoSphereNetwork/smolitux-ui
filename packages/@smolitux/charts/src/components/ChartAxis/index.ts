export { default, type ChartAxisProps, type ChartAxisTick } from './ChartAxis';
