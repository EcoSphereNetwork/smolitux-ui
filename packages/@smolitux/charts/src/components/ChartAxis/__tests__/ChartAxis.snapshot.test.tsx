import React from 'react';
import { render } from '@testing-library/react';
import { ChartAxis } from '../ChartAxis';

describe('ChartAxis Snapshot', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <svg>
        <ChartAxis length={100} ticks={[{ value: 0, position: 0 }]} />
      </svg>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
