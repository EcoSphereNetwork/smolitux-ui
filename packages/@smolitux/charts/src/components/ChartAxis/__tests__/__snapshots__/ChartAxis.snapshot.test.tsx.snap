// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartAxis Snapshot matches snapshot 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="smolitux-chart-axis"
      data-testid="chart-axis"
    >
      <line
        stroke="currentColor"
        x1="0"
        x2="100"
        y1="0"
        y2="0"
      />
      <g
        class="tick"
      >
        <line
          stroke="currentColor"
          x1="0"
          x2="0"
          y1="0"
          y2="6"
        />
        <text
          fill="currentColor"
          font-size="12"
          text-anchor="middle"
          x="0"
          y="18"
        >
          0
        </text>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
